src/main/kotlin/com/team4099/robot2023/subsystems/elevator/ElevatorIOSim.kt

```diff
@@ -0,0 +1,146 @@
+import org.team4099.lib.units.base.*
+import org.team4099.lib.units.derived.DerivativeGain
+import org.team4099.lib.units.derived.ElectricalPotential
+import org.team4099.lib.units.derived.IntegralGain
+import org.team4099.lib.units.derived.ProportionalGain
+import org.team4099.lib.units.derived.Volt
+import org.team4099.lib.units.derived.inVolts
+import org.team4099.lib.units.derived.volts
+import org.team4099.lib.units.perSecond
+
+object ElevatorIOSim : ElevatorIO {
+
+    val elevatorSim: ElevatorSim = ElevatorSim(
+            DCMotor.getNEO(2),
+            ElevatorConstants.GEAR_RATIO,
+            ElevatorConstants.CARRIAGE_MASS.inKilograms,
+            ElevatorConstants.SPOOL_DIAMETER.inMeters/2,
+            ElevatorConstants.ELEVATOR_MAX_RETRACTION.inMeters,
+            ElevatorConstants.ELEVATOR_MAX_EXTENSION.inMeters,
+            true,
+            0.0
+        )
+
+    init {
+        MotorChecker.add(
+            "Elevator",
+            "Extension",
+            MotorCollection(
+                mutableListOf(
+                    SimulatedMotor(
+                        elevatorSim,
+                        "Extension Motor",
+                    )
+                ),
+                baseCurrentLimit = 60.amps,
+                firstStageTemperatureLimit = 10.celsius,
+                firstStageCurrentLimit = 45.amps,
+                motorShutDownThreshold = 20.celsius
+            )
+        )
+    }
+
+    private val elevatorController =
+        PIDController(ElevatorConstants.SIM_KP, ElevatorConstants.SIM_KI, ElevatorConstants.SIM_KD)
+
+    private var lastAppliedVoltage = 0.0.volts
+
+    override fun updateInputs(inputs: ElevatorIO.ElevatorInputs) {
+        elevatorSim.update(Constants.Universal.LOOP_PERIOD_TIME.inSeconds)
+
+        inputs.elevatorPosition = elevatorSim.positionMeters.meters
+        inputs.elevatorVelocity = elevatorSim.velocityMetersPerSecond.meters.perSecond
+
+        inputs.leaderTempCelcius = 0.0.celsius
+        inputs.leaderStatorCurrent = 0.0.amps
+        inputs.leaderSupplyCurrent = elevatorSim.currentDrawAmps.amps / 2
+        inputs.leaderAppliedVoltage = lastAppliedVoltage
+
+        inputs.followerTempCelcius = 0.0.celsius
+        inputs.followerStatorCurrent = 0.0.amps
+        inputs.followerSupplyCurrent = elevatorSim.currentDrawAmps.amps / 2
+        inputs.followerAppliedVoltage = lastAppliedVoltage
+
+        inputs.leaderRawPosition = 0.0
+        inputs.followerRawPosition = 0.0
+
+        inputs.isSimulating = true
+
+        RoboRioSim.setVInVoltage(
+            BatterySim.calculateDefaultBatteryLoadedVoltage(elevatorSim.currentDrawAmps)
+        )
+    }
+
+    /**
+     * Sets the voltage of the elevator motors but also checks to make sure elevator doesn't exceed
+     * limit
+     *
+     * @param voltage the voltage to set the motor to
+     */
+    override fun setOutputVoltage(voltage: ElectricalPotential) {
+        val clampedVoltage =
```

Comment:
should wrap this in an if statement that checks whether our elevator is either
1. `<0.5 inches` and applying a negative voltage
2. `> max_height-0.5 inches` and applying a positive voltage

if either of these are true, we don't want to be able to run the motor bc that will break elevator